File: 프로그래머스/3/152995. 인사고과/인사고과.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

bool cmp(vector<int> v1,vector<int> v2){
    return v1[0]>v2[0];
}

int solution(vector<vector<int>> scores) {
    int answer = 1;
    int WonScore = scores[0][0] + scores[0][1];
    int s1 = scores[0][0];
    int s2 = scores[0][1];
    
    sort(scores.begin(), scores.end(), cmp);
    
    int a = scores[0][1], b = scores[0][1];
    for(int i = 0; i<scores.size(); i++){
        if(s1 < scores[i][0] and s2 < scores[i][1]) return -1;
        
        
        if(i == 0){
            if(WonScore < scores[i][0] + scores[i][1]) answer++;
        }
        else{
            if(scores[i - 1][0] >scores[i][0]) a = b;
                
            if(a > scores[i][1]) continue;
                
            if(WonScore < scores[i][0] + scores[i][1]) answer++;      
            b = max(scores[i][1], b);
        }
    
    }
    
    return answer;
}

File: 프로그래머스/3/152995. 인사고과/인사고과.java
import java.io.*;
import java.util.*;

class Solution {
    public int solution(int[][] scores) {
        int answer = 1;
        
        int a = scores[0][0]; int b = scores[0][1];
        
        int n = scores.length;
        xy[] v = new xy[n];
        
        for(int i = 0
            ; i < n; i++){
            v[i] = new xy();
            v[i].x = scores[i][0];
            v[i].y = scores[i][1];
        }
        
        Arrays.sort(v, (x, y) -> {
            if(x.x == y.x) return x.y - y.y;
            return y.x - x.x;
        });

        
        int max = v[0].y;
        for(int i = 0; i < n; i++){
            if(v[i].y < max){ 
                if(v[i].x == a && v[i].y == b){
                    return -1;
                }
                v[i].x = v[i].y = 0;
            }
            else max = Math.max(v[i].y , max);
        }
        
        for(int i = 0; i < n; i++){
            if(a + b < v[i].x + v[i].y) answer++;
        }
        
        
        return answer;
    }
    static class xy{
        int x, y;
    }
}